xwiki-platform-core/xwiki-platform-oldcore/src/main/java/org/xwiki/internal/migration/AbstractDocumentsMigration.java

```diff
@@ -65,28 +80,15 @@ public abstract class AbstractDocumentsMigration extends AbstractHibernateDataMi
     @Override
     protected void hibernateMigrate() throws DataMigrationException
     {
-        List<XWikiDocument> documents = selectDocuments()
-            .stream()
-            .map(this::convert)
-            .filter(Optional::isPresent)
-            .map(Optional::get)
-            .collect(toList());
-        logBeforeQueuingTasks(documents);
-        for (XWikiDocument document : documents) {
-            logBeforeQueuingTask(document);
-            this.taskManager.addTask(this.getXWikiContext().getWikiId(), document.getId(), getTaskType());
-        }
-    }
+        List<ReferenceWithLocale> selectedDocuments = selectDocuments();
+        logBeforeQueuingTasks(selectedDocuments);
 
-    private Optional<XWikiDocument> convert(String documentReference)
-    {
-        XWikiContext context = getXWikiContext();
-        try {
-            return Optional.of(
-                context.getWiki().getDocument(this.documentReferenceResolver.resolve(documentReference), context));
-        } catch (XWikiException e) {
-            this.logger.error("Failed to resolve [{}]", documentReference, e);
-            return Optional.empty();
+        for (ReferenceWithLocale referenceWithLocale : selectedDocuments) {
+            XWikiDocument document =
+                new XWikiDocument(this.documentReferenceResolver.resolve(referenceWithLocale.reference),
+                    toLocale(referenceWithLocale.locale, ROOT));
```

Comment:
Note that `toLocale` might fail here when the document in the database has an unsupported locale (it's unlikely, but it did happen not long ago) and from what I understand it would crash the rest of the migration.

Reply:
Can `toLocale` with a fallback fail?

Reply:
For instance, the following returns `true`

```java
assertEquals(Locale.CANADA, LocaleUtils.toLocale("aaaaaa", Locale.CANADA));
```

Reply:
Indeed, it cannot fail if you pass a fallback. But it does not make sense to fallback on ROOT, if that document has an unsupported locale it should be skipped.